packages/design-system/src/components/IdleTimeout/IdleTimeout.tsx

```diff
@@ -77,70 +79,146 @@ const defaultMessageFormatter = (timeTilTimeout: number): React.ReactNode => {
-      onSessionContinue();
+}: IdleTimeoutProps): JSX.Element => {
+  if (timeToWarning > timeToTimeout) {
+    console.error(
+      'Error in TimeoutManager component. `timeToWarning` is greater or equal to `timeToTimeout`'
+    );
+  }
+  const MS_BETWEEN_STATUS_CHECKS = 30000;
+  // convert minutes to milliseconds
+  const msToTimeout = timeToTimeout * 60000;
+  const msToWarning = (timeToTimeout - timeToWarning) * 60000;
+  const [checkStatusTime, setCheckStatusTime] = useState<number>(null);
+  const [showWarning, setShowWarning] = useState<boolean>(false);
+  const [timeInWarning, setTimeInWarning] = useState<number>(timeToWarning);
+
+  // cleanup timeouts & intervals
+  const clearTimeouts = () => {
+    setCheckStatusTime(null);
+  };
+
+  // when the countdown for the session ends, clean up, call callback & close modal
+  const handleTimeout = () => {
+    clearTimeouts();
+    removeEventListeners();
+    onTimeout();
+    setShowWarning(false);
+  };
+
+  // when it's time to warn the user about idleness,
+  // set an interval that updates the modal message
+  const handleWarningTimeout = () => {
+    setShowWarning(true);
+    removeEventListeners();
+  };
+
+  const setTimeoutCookies = () => {
+    const timeoutTime = Date.now() + msToTimeout;
+    const warningTime = Date.now() + msToWarning;
+    localStorage.setItem(timeoutCookieName, timeoutTime.toString());
+    localStorage.setItem(timeoutWarningCookieName, warningTime.toString());
+
+    if (checkStatusTime === null) {
+      setCheckStatusTime(MS_BETWEEN_STATUS_CHECKS);
     }
   };
 
-  const handleEndSession = () => {
-    if (onSessionForcedEnd) {
-      // TODO: figure out if any params should be passed back to the app
-      onSessionForcedEnd();
-    } else if (onTimeout) {
-      onTimeout();
+  const resetTimeouts = () => {
+    clearTimeouts();
+    setTimeoutCookies();
+  };
+
+  const removeEventListeners = () => {
+    document.removeEventListener('mousemove', resetTimeouts);
+    document.removeEventListener('keypress', resetTimeouts);
+  };
+
+  const addEventListeners = () => {
+    const passiveSupported = checkPassiveSupport();
+    const options = passiveSupported ? { passive: true } : false;
+    document.addEventListener('mousemove', resetTimeouts, options);
+    document.addEventListener('keypress', resetTimeouts, options);
+  };
+
+  const checkWarningStatus = () => {
+    const warningTime = Number(localStorage.getItem(timeoutWarningCookieName));
+    const timeoutTime = Number(localStorage.getItem(timeoutCookieName));
+    const now = Date.now();
+
+    if (now >= Number(timeoutTime)) {
+      handleTimeout();
+    } else if (!showWarning && now >= warningTime) {
+      handleWarningTimeout();
+    } else if (showWarning && now >= warningTime) {
+      // if the warning is showing, update the timeInWarning variable (in minutes)
+      const minutesLeft = Math.ceil((timeoutTime - now) / 60000);
+      setTimeInWarning(minutesLeft);
```

Comment:
Doing it in this block may mean we don't have an initial value after `handleWarningTimeout()`, if I'm thinking about it correctly. If we moved this `minutesLeft` calculation into the main function body and provided it directly to the dialog message format functiion, we wouldn't even need to store `timeInWarning` in local state. I could be wrong

Reply:
Another thing I just thought about in this branching logic here: Do we handle the case where another tab updates the `localStorage` values. Is there a way this component will read an updated value and _hide_ the warning modal? That is, this tab is starting to time out, but in another tab the end user is working away and updating `localStorage` with fresh timeout times. I haven't actually tested it yet in my browser

Reply:
>Doing it in this block may mean we don't have an initial value after handleWarningTimeout(), if I'm thinking about it correctly. If we moved this minutesLeft calculation into the main function body and provided it directly to the dialog message format functiion, we wouldn't even need to store timeInWarning in local state. I could be wrong

There is an initial value after `handleWarningTimeout` because the state variable is initialized with a value (`timeToWarning`)

Aside from not having a state variable, I'm not sure how this change would be much different. As I'm implementing it, in order to have all the information to calculate minutes left, I would have to do similar calculations in another function that this function is already doing and wouldn't it be performing the calculation for `minutesLeft` on every render?